server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesAggregator.java

```diff
@@ -0,0 +1,123 @@
+        AggregatorFactories factories,
+        boolean keyed,
+        AggregationContext context,
+        Aggregator parent,
+        CardinalityUpperBound bucketCardinality,
+        Map<String, Object> metadata
+    ) throws IOException {
+        super(name, factories, context, parent, bucketCardinality, metadata);
+        this.keyed = keyed;
+        tsidFieldData = (IndexFieldData<SortedSetBytesLeafFieldData>) Objects.requireNonNull(
+            context.buildFieldContext("_tsid"),
+            "Cannot obtain tsid field"
+        ).indexFieldData();
+        bucketOrds = BytesKeyedBucketOrds.build(bigArrays(), bucketCardinality);
+    }
+
+    @Override
+    public InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
+        InternalTimeSeries.InternalBucket[][] allBucketsPerOrd = new InternalTimeSeries.InternalBucket[owningBucketOrds.length][];
+        for (int ordIdx = 0; ordIdx < owningBucketOrds.length; ordIdx++) {
+            BytesRef spareKey = new BytesRef();
+            BytesKeyedBucketOrds.BucketOrdsEnum ordsEnum = bucketOrds.ordsEnum(owningBucketOrds[ordIdx]);
+            List<InternalTimeSeries.InternalBucket> buckets = new ArrayList<>();
+            while (ordsEnum.next()) {
+                long docCount = bucketDocCount(ordsEnum.ord());
+                ordsEnum.readValue(spareKey);
+                InternalTimeSeries.InternalBucket bucket = new InternalTimeSeries.InternalBucket(
+                    TimeSeriesIdFieldMapper.decodeTsid(spareKey),
+                    docCount,
+                    null,
+                    keyed
+                );
+                bucket.bucketOrd = ordsEnum.ord();
+                buckets.add(bucket);
+            }
+            allBucketsPerOrd[ordIdx] = buckets.toArray(new InternalTimeSeries.InternalBucket[0]);
+        }
+        buildSubAggsForAllBuckets(allBucketsPerOrd, b -> b.bucketOrd, (b, a) -> b.aggregations = a);
+
+        InternalAggregation[] result = new InternalAggregation[owningBucketOrds.length];
+        for (int ordIdx = 0; ordIdx < owningBucketOrds.length; ordIdx++) {
+            result[ordIdx] = buildResult(allBucketsPerOrd[ordIdx]);
+        }
+        return result;
+    }
+
+    @Override
+    public InternalAggregation buildEmptyAggregation() {
+        return new InternalTimeSeries(name, new ArrayList<>(), false, metadata());
+    }
+
+    @Override
+    protected void doClose() {
+        Releasables.close(bucketOrds);
+    }
+
+    @Override
+    protected LeafBucketCollector getLeafCollector(LeafReaderContext context, LeafBucketCollector sub) throws IOException {
+        final SortedBinaryDocValues tsids = tsidFieldData.load(context).getBytesValues();
+        return new LeafBucketCollectorBase(sub, null) {
+
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                if (tsids.advanceExact(doc)) {
+                    BytesRef newTsid = tsids.nextValue();
```

Comment:
So, I would like to replace this whole part in a follow up with getting data based on the actually specified index sort order instead of hard coded tsids and timestamps. Are you ok, if I address these optimizations there?

Reply:
Totally fine with a follow-up.